controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -237,4 +245,104 @@ public boolean isDynamicDependent() {
+            assert attributeValues.length <= 1; //we can only handle single attribute value
+            super.processCapabilityRequirement(context, resource,attributeName, remove, attributeValues);
+            RuntimeCapability cap = getDependentCapability(context);
+            String dependentName = getDependentName(cap, context);
+            boolean optional = cap.getDynamicOptionalRequirements().contains(baseRequirementName);
+            for (String attributeValue : attributeValues) {
+                if (attributeValue != null || !optional) {
+                    String requirementName = getRequirementName(context, resource.getModel(), attributeValue);
+                    if (remove) {
+                        context.deregisterCapabilityRequirement(requirementName, dependentName);
+                    } else {
+                        context.registerAdditionalCapabilityRequirement(requirementName, dependentName, attributeName);
+                    }
+                }
+            }
+        }*/
+
+        @Override
+        protected RuntimeCapability getDependentCapability(OperationContext context) {
+            if (capability != null) {
+                return capability;
+            }
+            return super.getDependentCapability(context);
+        }
+
+        @Override
+        protected String getRequirementName(OperationContext context, String attributeValue) {
+            ModelNode model = context.readResource(PathAddress.EMPTY_ADDRESS).getModel();
+            String[] dynamicParts = new String[attributes.length + 1];
+            try {
+                for (int i = 0; i < attributes.length; i++) {
+                    SimpleAttributeDefinition ad = attributes[i];
+                    dynamicParts[i] = ad.resolveModelAttribute(context, model).asString();
+                }
+            } catch (OperationFailedException e) {
+                throw new RuntimeException(e);
+            }
+            dynamicParts[attributes.length] = attributeValue;
+            return RuntimeCapability.buildDynamicCapabilityName(baseRequirementName, dynamicParts);
+        }
+
+    }
+
+    /**
+     * {@link ResourceNameCompositeDependencyRecorder} that determines the dependent capability
+     * from the {@link OperationContext} and resource name.
+     * This assumes that the {@link OperationContext#getResourceRegistration() resource registration associated with currently executing step}
+     * will expose a {@link ImmutableManagementResourceRegistration#getCapabilities() capability set} including
+     * one and only one capability. <strong>This recorder cannot be used with attributes associated with resources
+     * that do not meet this requirement.</strong>
+     */
+    public class ResourceNameCompositeDependencyRecorder extends CompositeAttributeDependencyRecorder {
```

Comment:
This one is actually needed as this recorder is not used in builders.
Maybe it would make sense to add new method to attribute builder that would use this class as well.